Create watershed API

rod-glover opened the pull request:
This PR adds a new endpoint, `/streamflow/watershed` to the CE backend API.

This PR also breaks up the monolithic file `tests/test_api.py` into test files related to each individual API endpoint under new directory `tests/api/`. Tests for this particular API are in `tests/api/streamflow/test_watershed.py`, and are the only significant changes. The rest are existing tests, moved.

See the documentation in `api/streamflow/watershed.py` for a description of the API.

A test version of this API has been stood up at `http://docker-dev01.pcic.uvic.ca:30031/api/streamflow/watershed`. This version is backed by a test modelmeta database also running on docker-dev01.

For example, try [this URL](http://docker-dev01.pcic.uvic.ca:30031/api/streamflow/watershed?ensemble_name=hydro_files&station=POINT(-118%2054.7)). Feel free to vary the value of `station`. Note there is no comma between POINT coordinates.

@jameshiebert , do you know how/where the decision was made that the hypsometry histogram should have `sqrt(len(cells))` bins, and that the range of elevations (yielding bin sizes) should be padded by 5 elevation units? I might propose that the bins width, at least, be rounded to a suitable multiple of `10^n`. Or something like that, rather than the unrounded value we use now.

rod-glover commented:
@corviday , some questions/suggestions. It is quite possible that these are naive and/or settled this way for some reasons not clear to me.

URL:

1. What is the reasoning behind the two-level URL `/streamflow/watershed` ? Could it reasonably be just one or the other?

1. Is param `ensemble_name` necessary?

Resource returned (see above):

1. Suggestion: Factor out common info in elevation props, thus:

   ```
   "elevation": {
     "units": "m",
     "minimum":  797.2999877929688,
     "maximum": 837.0999755859375
   }
   ```

1. In prop "hyspometric_curve", the "histogram" prop seems unnecessary. Can it reasonably be simplified to
   ```
   "hyspometric_curve": {
     "bin_width": 24.899993896484375,
       "x_bin_centers": [
         804.7499847412109,
         829.6499786376953
       ],
       "y_values": [
         2,
         1
       ],
      "x_units": "m",
      "y_units": "grid cells"
   }
   ```

1. Do we need CRS spec (WGS84 would likely be it) in the `shape` component? Maybe that is a question for MoTI.

corviday commented:
> 1. What is the reasoning behind the two-level URL /streamflow/watershed ? Could it reasonably be just one or the other?

I vaguely envisioned this endpoint as fitting into our design for the routed streamflow API, where we planned to have `/streamflow/orders` and `/streamflow/results` endpoints. My thought was that a `/streamflow/results/id` response would link to the corresponding `/streamflow/watershed`, which would provide contextual information for the streamflow data, like the size of the watershed, when we got that far, and maybe a `/streamflow/watershed` would like to already-calculated or in-process data for that cell. 

I'm not attached to the idea; we could reasonably just name it `watershed`.

> 2. Is param ensemble_name necessary?

It's just the name of the ensemble the datafiles are in; it seemed like the best way to find them.

> 3. Do we need CRS spec (WGS84 would likely be it) in the shape component? Maybe that is a question for MoTI.

The shape component was not requested by MOTI; I added it for debugging (though MOTI will probably like it, and if it doesn't slow things too much and they like it we should leave it in for them). So we could ask them their preferences, but they may not have any.

I looked to see if we could just get a CRS from the datafiles, but none of them have one.

corviday commented:
I've now made all the changes I was planning to:

- add a check to ensure that area, elevation, and flow datasets are all using the same grid
- format improvements
- switching the hypsometric data from "grid cells per elevation" to "square meters per elevation".

I'll update the demo, and then it's all yours.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,348 @@
+        neighbour_routing = routing[neighbour]
+        if neighbour_routing is np.ma.masked:
+            return False
+        # `neighbour` is upstream if its routing points back at `cell`
```
Tighter:

```
try:
    neighbour_routing = int(routing[neighbour])
except IndexError, <numpy error>:
    return False
```

And lose `is_valid_index`.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,348 @@
+                    'longitude': lon,
+                    'latitude': lat
+                },
+                'area': outline.area,
```
Remove

rod-glover commented on ce/api/util.py:
```diff
@@ -153,3 +155,32 @@ def search_for_unique_ids(sesh, ensemble_name='ce', model='', emission='',
+    return set(product(range(*tuplify(m)), range(*tuplify(n))))
+
+
+def is_valid_index(index, shape):
```
Remove

rod-glover commented on ce/api/util.py:
```diff
@@ -153,3 +155,32 @@ def search_for_unique_ids(sesh, ensemble_name='ce', model='', emission='',
+
+
+def vec_add(a, b):
+    '''numpy-style addition for builtin tuples: (1,1)+(2,3) = (3,4)'''
```
Double quotes

rod-glover commented:
@jameshiebert , @nikola-rados , hold off on the review. I am making some changes to make the main `watershed` method testable.

rod-glover commented:
TO DO:

- [ ] Respond with 404 Not Found, not 500 Internal Error, when the specified station location is not within the watershed.

rod-glover commented:
An up-to-date demo of this code is running on docker-dev01: [example](http://docker-dev01.pcic.uvic.ca:30031/api/streamflow/watershed?ensemble_name=hydro_files&station=POINT(-117%2054.7))

jameshiebert commented:
Do you still want us to hold off on review until your TODO list is checked off? Or are you saying that those things can be deferred to a new PR?

And are you going to fix the CI tests?

Can't wait to read through this; it looks like great work!

rod-glover commented:
@jameshiebert , @nikola-rados , please do hold off on reviewing still. I will ping you when I'm ready. At least one of the TODOs have to be addressed before I am happy releasing this, and I don't want you to waste your effort on unfinished code.

rod-glover commented:
@jameshiebert , @nikola-rados , I'm ready for your review. Thanks for your patience. And once again, that URL is http://docker-dev01.pcic.uvic.ca:30031/api/streamflow/watershed?ensemble_name=hydro_files&station=POINT(-117%2054.7) . Call now! :)

rod-glover commented:
A few notes to guide the reviewers:

- This PR is substantial but it's not as big as it looks. Quite a bit of the "new" code is just tests moved from one humongous test file to individual ones. There's also a little code reformatting, mostly a matter of spacing and line lengths.
- You can ignore all test files except those under `ce/tests/api/streamflow/watershed`.
- The place to start is `ce/api/streamflow/watershed.py`. GitHub thinks it's long, so it's rolled up and easy to miss.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+        switching the order of the coordinates. Checks that the index is
+        valid for the grid; we must at minimum exclude negative index values,
+        which are valid but wrong in this application."""
+        x = int(round((lonlat[1] - self.latitudes[0]) / self.lat_step))
```
Maybe that datasets that will be used for the watershed API *are* consistent. But I would recommend at least an `assert` statement ensuring that the deltas are all equal, if the rest of the code depends on that.

rod-glover replied:
Hmm, this is a trickier question than it at first seems. It's worth a little analysis:

- The watershed (upstream cells from mouth) is built based in part on the sign of the lat an lon steps (this determines the "direction matrix"). But this is not a compatibility issue.
- Upstream cells are returned as cell indices. They are converted to lonlat pairs for the next steps.
- The elevation and area lists are built by querying on lonlats (converted from upstream cell indices). I don't *believe* this code requires even that elevation and area datasets are on the same grid: For each variable (elevation, area), a lonlat is converted to the nearest cell index and that is used. 
   - Now the question is raised of whether this approximation is acceptable, or whether we must insist that the grids are compatible in the sense that each lonlat specified corresponds exactly to a lonlat in the grid.
- Therefore the comparison of whether lonlat grids are compatible because they have the same step is either
    - irrrelevant, or
    - insufficient; we need to check whether the lonlats are the same points, not just the same distance apart.

@corviday , do you have any thoughts on this?

rod-glover replied:
Some further thoughts:

- I imagined the purpose of `DataGrid`, despite its general sounding name, to be to handle only the flow direction, elevation, and area datasets needed in the watershed API. So at present it builds in some assumptions about those datasets inferred from the examples we have. Those assumptions are:
    - Uniformly spaced latitude and longitude grids
    - A loose notion of compatibility between grids, which looks like it probably was too loose (see above)

- If we want `DataGrid` to represent more general datasets, then it will have to drop these assumptions. Furthermore, it probably wouldn't be appropriate to build in an `is_compatible` method, since the compatibility criterion could vary greatly from application to application. Little as I like object oriented code in general, this would be a case for a base class with usage specific subclasses.

- So maybe it needs (a) a different name that implies its specific usage, and (b) subclassing from a more generic class.

rod-glover replied:
I refactored DataGrid into a base and VIC/watershed-specific subclass, and added a more stringent compatibility condition, which is that the lat and lon coordinates must be the same values where they overlap spatially.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+            outline,
+            properties={
+                # Represent as GeoJSON?
+                'mouth': {
```
Should we consider "snapping" the mouth the the center of the grid cell in which the input mouth falls?

rod-glover replied:
Good idea. Any thought on whether this should be a GeoJSON representation?

rod-glover replied:
Actually, I prefer GeoJSON -- any reason why not?

jameshiebert replied:
GeoJSON sounds great.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+    """
+    station_lonlat = WKT_point_to_lonlat(station)
+
+    with get_time_invariant_variable_dataset(
```
I think that you can specify multiple context managers in a single with statement to avoid the massive indent. I.e.

```
with foo() as bar, blah() as stuff:
    do_something_with(bar, stuff)
```

rod-glover replied:
Oh, man, I was wishing for that.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+                        area=DataGrid.from_nc_dataset(area_ds, 'area'),
+                    )
+                except DataGridIndexError:
+                    abort(404, description=
```
That seems like a very sensible way to handle this error input condition. Nice.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+    """
+    if hypso_params is None:
+        # Default parameters cover total range of BC elevations from
+        # sea level to highest point, Mt. Fairweather, at 4,663 m.
```
Fine for now, but [Mount Logan](https://en.wikipedia.org/wiki/Mount_Logan) isn't far afield of BC's borders. Flag this if we start modelling the coast range. Rainier and Hood are within this range, so we're probably good for the Columbia.

jameshiebert commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+        },
+        'area': {
+            'units': area.units,
+            'value': sum(ws_areas),
```
I had envisioned having `shapely` compute the area from the outline, but I guess this works too.

rod-glover commented on ce/api/geospatial.py:
```diff
@@ -0,0 +1,53 @@
+    ])
+
+
+def outline_point_buff(centres, height, width, resolution=16):
```
It's a leftover, will delete.

rod-glover commented on ce/api/geospatial.py:
```diff
@@ -0,0 +1,53 @@
+    return cascaded_union(cells)
+
+
+def WKT_point_to_lonlat(text):
```
Good idea. And prompted improved error handling too. yay.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+arranged in lat-lon order.
+
+Spatial coordinates are named lat and lon. Data indexes are named x and y,
+and correspond to lat and lon.
```
Nope, but I better clarify this remark.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Set of cells (cell indices) that drain into `target`.
```
Only if we are comfortable with requiring that flow_direction, elevation, and area datasets are on *exactly* the same grid, so that the indices in each for a given latlon are identical. The mapping back and forth makes it robust to differences in grid, which might be a good thing, or might not. See also my question in response to comment on compatibility tests.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+        )
+
+    if debug:
+        return upstream(target), {'time': time.time() - start_time}
```
Definitely: that was a rough bit of coding. And a very nice alternative for timing! I learn something every day.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
```
No, this is the N/S/E/W flow direction to index offset map, which we do compute in this application. It's just a question of where the call should be.

rod-glover commented on ce/api/streamflow/watershed.py:
```diff
@@ -0,0 +1,421 @@
+    return Dataset(file.filename, 'r')
+
+
+def hypsometry(elevations, areas, bin_start=0, bin_width=100, num_bins=46):
```
I don't think this indicates a problem in either the data or the code. 

In the elevation file we have (presumably from HI), we see 

```
$ ncdump -h elevation.nc 
netcdf elevation {
dimensions:
	lon = 1088 ;
	lat = 512 ;
variables:
	double lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "longitude" ;
		lon:units = "degrees_east" ;
		lon:axis = "X" ;
	double lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "latitude" ;
		lat:units = "degrees_north" ;
		lat:axis = "Y" ;
	float elev(lat, lon) ;
		elev:long_name = "Average elevation of grid cell" ;
		elev:units = "m" ;
		elev:_FillValue = NaNf ;
		elev:missing_value = NaNf ;
		elev:cell_methods = "area: mean" ;
```

I'm taking `elev:long_name = "Average elevation of grid cell" ` at face value. In which case, skipping elevation bands is not contradictory: Consider a case where the area in question contains lowlands and highlands with very steep changes between them. Then the set of average elevations could easily jump from, say, 100 m to 500 m, missing the intervening elevation bands. This will happen wherever the topography varies wildly. We have a lot of that in BC.

rod-glover replied:
This is much more likely to occur when there are just a few cells in the watershed, since `n` cells can represent at most `n` elevations, and the possibility that there is a steep area whose missing elevation bands are not filled in by some other area is much greater. The example you give has 9 cells, so I'd conclude that this is exactly what we might expect.

rod-glover commented on ce/api/util.py:
```diff
@@ -153,3 +166,31 @@ def search_for_unique_ids(sesh, ensemble_name='ce', model='', emission='',
     return ( r[0] for r in query.all() )
+
+
+def index_set(m, n):
```
Not really. This returns index pairs; `meshgrid` returns something related but quite different.

rod-glover commented on ce/api/util.py:
```diff
@@ -153,3 +166,31 @@ def search_for_unique_ids(sesh, ensemble_name='ce', model='', emission='',
+    return all(0 <= i < n for i, n in zip(index, shape))
+
+
+def vec_add(a, b):
```
For better or worse, not all the code in this PR is `numpy` oriented. Possibly it should be. I'll put in an issue but leave this as is.

rod-glover commented:
I *think* I have addressed all the feedback, with code and/or answers to the comments. The major items, in brief:

- Assumptions about the spatial grids of datasets, specifically flow_direction, elevation, area:
  - Those datasets do have uniformly spaced grids.
  - The flow direction, elevation, and area datasets have coincident grids, which means they have the same absolute step size and where they overlap spatially they are on the same points
  - The spatial extent of flow direction is different from that of elevation and area
  - The spatial extent of elevation and area datasets are the same (as they ought to be)

- "Missing elevations" in the hypsometry are not erroneous. Especially in small watersheds, it is easy for steep terrain with widely varying elevations to show gaps between elevation bins. For example, lowlands at 100 m elevation and highlands at 500 m, with no intervening bins because there are no cells in the small watershed at those elevations. Steep terrain does that.

rod-glover commented:
Modified code is running on docker-dev01

jameshiebert commented:
Looks good @rod-glover , thanks for addressing the comments in short order! I'm happy if you go ahead and merge this. I think there are a couple outstanding issues (for which we should create [new issues](https://github.com/pacificclimate/climate-explorer-backend/issues/new):

- [ ] Upgrade the TravisCI base image to `bionic` and test against py37
- [ ] Write user docs to explain the possible gaps in hypsometric curve

I agree that the gaps in the curve are physically possible and sensible if you think about the various scale mismatches. But it will be necessary to write this up for users of the API, so that they can expect it as well.
